wp/lib/bap_wp/src/precondition.ml

```diff
@@ -1289,16 +1289,39 @@ let user_func_spec
   debug "Making user-defined subroutine spec with subroutine-name: %s, pre:
 %s, post: %s \n%!" sub_name sub_pre sub_post;
   if String.equal sub_name (Sub.name sub) then
-    let spec env post tid = (
+    let spec env post _ = (
       (* turn strings into proper smtlib2 statements; incr stack_ptr *)
-      let sub_pre : Constr.t = Z3_utils.mk_smtlib2_single env sub_pre in
-      let sub_post : Constr.t = Z3_utils.mk_smtlib2_single env sub_post in
+      let (sub_pre, pre_init_vars, pre_vars)
+          : Constr.t  * string list * string list =
+        Z3_utils.mk_smtlib2_single_with_vars env sub_pre in
+      let (sub_post, post_init_vars, post_vars)
+          : Constr.t * string list * string list =
+        Z3_utils.mk_smtlib2_single_with_vars env sub_post in
       let sub_post, env = increment_stack_ptr sub_post env in
-      (* collect (physical) inputs/outputs of sub *)
-      let sub_inputs : Var.t list = vars_from_sub env sub |> Var.Set.to_list in
-      let sub_inputs =
-        List.filter sub_inputs ~f:(fun v -> Var.is_physical v) in
-      let sub_outputs : Var.t list = sub_inputs in
+      (* TODO: Some sanity checking on sub_pre and sub_post would be nice
+         here.  If you just forget the word "assert" you silently get some
+         kind of empty constr. *)
+
+      (* Next, collect inputs/outputs of sub.  We do this by constructing a set
+         of all potential inputs/outputs (all the registers and mem), then
+         filtering out everything not mentioned by the user-provided specs.
+
+         Question: is SP + GPRs really all the registers?  *)
+      let potential_ios : Bap.Std.Var.Set.t =
+        Set.add (Set.add (Env.get_gprs env) (Env.get_sp env)) (Env.get_mem env)
+      in
+      let filter_out_unused (used : string list) (adjust : string -> string) =
+        Set.filter potential_ios
+          ~f:(fun v ->
+            List.exists used
+              ~f:(fun s -> String.equal (adjust (Var.to_string v)) s))
+      in
+      let (sub_inputs, sub_outputs : Var.t list * Var.t list) =
+        (Set.to_list (filter_out_unused (pre_init_vars @ post_init_vars)
+                        (fun s -> "init_" ^ s)),
```

Comment:
We might want to clean the name before adding the `init_` like we do over here: https://github.com/draperlaboratory/cbat_tools/blob/7bb7e36b247860d81f8b4ce43779c1ceb1b76547/wp/lib/bap_wp/src/environment.ml#L443

Reply:
I think this is not needed.  We want to check whether the user has written _exactly_ one of the register/memory names. right?  Like `RAX` or `init_RAX`.  If we cleaned the name wouldn't we risk mistaking something else the user wrote for a register name or memory?

Perhaps I'm missing some subtlety though.